lib/Doctrine/ORM/Id/LegacyIdGeneratorProxy.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Id;
+
+use Doctrine\Deprecations\Deprecation;
+use Doctrine\ORM\EntityManager;
+
+use function get_debug_type;
+
+/**
+ * @internal This trait will be removed in 3.0.
+ */
+trait LegacyIdGeneratorProxy
```

Comment:
That would generate an infinite recursion (`generateId()` and `generate()` would call each other) if you forget to override either method. That's why I moved our internal implementation into a trait.

Reply:
not sure thats something we need to account for, everyone will have a `generate` method already, not having both seems something only new code could achieve. Can't help everyone from shooting themselves in the foot.

Reply:
No, but we don't need to hand them the gun. 🙂

Reply:
I you see a way of moving this logic into the abstract class without having it default to an endless recursion, I'd happily implement it. Otherwise, I really believe we're better of this way.

Reply:
You hand them the gun anyways, you can easily get infinte recursions by doing flush in an UoW event for example. 

A simple workaround would be a private property that tracks you have already delegated `alreadyDelegatedToGenerate` and check that before calling the delegate function again, throwing if already true, otherwise setting it to true.

Reply:
Good idea. Let's do it that way.